File: backend/src/lib/lemma/workflows.py
import json
from datetime import datetime
from .datastore import save_application
from .docstore import save_tailored_resume

async def trigger_application_workflow(user_email: str, payload):
    try:
        # Import the AI service function inside the trigger function to avoid circular dependency
        from ...services.ai_service import generate_application_analysis
        
        jd_text = payload.get("jdText")
        resume_text = payload.get("resumeText")
        company_name = payload.get("companyName")
        role_name = payload.get("roleName")
        
        print(f"[Lemma Workflows] Starting workflow for {role_name or 'Unknown Role'} at {company_name or 'Unknown Company'} for {user_email}")
        
        # 2. Run AI reasoning
        analysis_result = await generate_application_analysis({
            "jdText": jd_text,
            "resumeText": resume_text,
            "companyName": company_name,
            "roleName": role_name
        })
        
        # 3. Construct application entry
        signal_score = analysis_result.get("signalScore") or {}
        app_entry = {
            "id": f"app_{int(datetime.utcnow().timestamp() * 1000)}",
            "company": analysis_result.get("company") or company_name or "Unknown Company",
            "role": analysis_result.get("role") or role_name or "Software Engineer",
            "status": "Applied",  # default Kanban column
            "dateApplied": datetime.utcnow().isoformat() + "Z",
            "matchScore": signal_score.get("fitScore", 50),
            "effort": signal_score.get("effort", "Medium"),
            "flag": signal_score.get("flag", "Green"),
            "flagReason": signal_score.get("flagReason", ""),
            "jdText": jd_text,
            "tailoredBullets": json.dumps(analysis_result.get("tailoredBullets", [])),
            "outreachMessages": analysis_result.get("outreachMessages") or {"confident": "", "curious": "", "concise": ""},
            "learningResources": analysis_result.get("learningResources") or [],
            "gaps": analysis_result.get("gaps") or [],
            "nudge3Dismissed": False,
            "nudge7Dismissed": False
        }
        
        # 4. Save to DataStore & DocStore
        saved_app = save_application(user_email, app_entry)
        if saved_app:
            save_tailored_resume(user_email, saved_app["id"], app_entry["tailoredBullets"])
            
        return saved_app
    except Exception as e:
        print(f"[Lemma Workflows] Workflow execution failed: {e}")
        raise e

File: frontend/src/lib/lemma/workflows.js
import { saveApplication } from './datastore';
import { saveTailoredResume } from './docstore';

/**
 * Orchestrates the full agentic job application workflow:
 * 1. Takes the parsed Job Description and the Candidate's Resume.
 * 2. Triggers the AI logic to extract details, score the fit, tailor the resume, draft outreach messages, and generate prep packs.
 * 3. Saves the compiled application to the datastore (Kanban pipeline).
 */
export async function triggerApplicationWorkflow({ jdText, resumeText, companyName, roleName, aiProvider }) {
  try {
    // 1. Call our Next.js API route that handles the AI reasoning chain
    // (We will call the API relative to the server host or directly run the reasoning helper function)
    
    // In a Next.js server context, we can import and run the reasoning logic directly 
    // to avoid HTTP overhead. We'll import it from our API routes / services layer.
    const { generateApplicationAnalysis } = await import('../../services/aiService');
    
    console.log(`Starting Workflow for: ${roleName || 'Unknown Role'} at ${companyName || 'Unknown Company'}`);
    
    // 2. Run the AI reasoning engines (Groq + Gemini)
    const analysisResult = await generateApplicationAnalysis({
      jdText,
      resumeText,
      companyName,
      roleName
    });

    // 3. Construct the application entry
    const appEntry = {
      id: `app_${Date.now()}`,
      company: analysisResult.company || companyName || 'Unknown Company',
      role: analysisResult.role || roleName || 'Software Engineer',
      status: 'Applied', // Default status in Kanban
      dateApplied: new Date().toISOString(),
      matchScore: analysisResult.signalScore.fitScore || 50,
      effort: analysisResult.signalScore.effort || 'Medium',
      flag: analysisResult.signalScore.flag || 'Green',
      flagReason: analysisResult.signalScore.flagReason || '',
      jdText: jdText,
      tailoredBullets: JSON.stringify(analysisResult.tailoredBullets || []),
      outreachMessages: analysisResult.outreachMessages || { confident: '', curious: '', concise: '' },
      interviewQuestions: analysisResult.interviewQuestions || [],
      gaps: analysisResult.gaps || [],
    };

    // 4. Save to Lemma DataStore & DocStore
    const savedApp = saveApplication(appEntry);
    if (savedApp) {
      saveTailoredResume(savedApp.id, appEntry.tailoredBullets);
    }

    return savedApp;
  } catch (error) {
    console.error('Lemma Workflow Engine Error:', error);
    throw error;
  }
}
